### DISCUSSION TITLE
BERTopic - Dealing with complaints

### DISCUSSION THREAD

**Discussion by PDuarteLopes** [MaartenGr/BERTopic#2171] [Q&A]
Hey all, 
I am working with a Topic Modelling project which has the goal of categorizing user complaints. Tough BERTopic is a really good model I am struggling due to the nature of having a lot of noise even with some pre-process cleaning (removal internal communication, acronyms and user data anonymization). Is there anything I can do to improve results, make my model deal better with noisy text and make my topics be more specific and concise? I am struggling when creating a lot of meaningless topics and missing the key aspects which make a topic a topic

**Comment by MaartenGr** [comment]
Thanks for reaching out! Could you give a couple of examples perhaps as to the behavior of BERTopic that you want improved? Without output (or training code) it's very difficult to say whether something is due to the code, algorithm, or perhaps even the data.

**Reply by PDuarteLopes** [reply]
Thanks for the quick response. I would like to know if it's possible to iteratively introduce a manually educated bias into the BERTopic model. I’m conducting several experiments on a set of complaints to see which topics emerge from them. However, after each iteration of training the model, I analyze the topics and conclude which ones are definitely correct and which need further tuning. In the next experiment, I would like to provide BERTopic with the information that certain topics, along with their top words, will definitely exist, helping it to better define the other topics. Is something like this possible with the available tools? 
Also beside cosine_similairty do you suggest other metric to see if two or more topics should be merged ? 

Best regards

**Reply by MaartenGr** [reply]
Apologies for the late reply! 

If you already have a set of topics that you think are correct, then you could use [zero-shot topic modeling](https://maartengr.github.io/BERTopic/getting_started/zeroshot/zeroshot.html) which allows you to specify a (sub-)set of topics that you have already found. 

Another solution would be to use semi-supervised topic modeling but that will not guarantee the topics you had already found.

**Reply by PDuarteLopes** [reply]
No problem, and thanks once again for your reply.

If I have only a set of isolated words/expressions, should I use zero-shot learning or seed topics and seed words? I am confident that some topics will emerge from my data, but I still want BERTopic to provide information on other possible topics that I might not have considered. I’d like to introduce a bias to make my model gravitate more toward certain words/expressions. Would you consider this approach semi-supervised topic modeling, and is it possible to achieve this?

Best regards,

**Reply by MaartenGr** [reply]
In that case, I would advise zero-shot as that provides topics you already found along with potential new ones. If you want the model to gravitate towards certain words/expressions you can also thrown in [seed words](https://maartengr.github.io/BERTopic/getting_started/seed_words/seed_words.html) as that does not affect the creation of the topics but merely their representations.

**Reply by PDuarteLopes** [reply]
Taking into consideration your responses, Zero-Shot appears to be more suitable for my project. I want the words and expressions I use as an "educated bias" to influence the model during the clustering phase, not just in representation. However, when applying my chosen personalized topics, the model does not converge as well as the previous "non-biased" model did, resulting in different clusters.

I'm now questioning if I'm using the best topic labels with my Zero-Shot approach. Do you have any advice on how to best formalize the topic labels?

I read the relevant documentation, and I understand that when a label is provided in Zero-Shot Topic Modeling, the model will first assign documents to the known topics before exploring new topics for any unassigned documents. Since the label is embedded, I assume that a precise topic label should yield better outcomes than a more general one. For example, should I use a label in a specific format like "sms"_resent_confirmation" or a more general title like "SMS Confirmation"?

-> Example:.
zeroshot_topic_list = [
    "Meter Replacement",                   # For ["replacement", "meter"]
    "Scheduling and Rescheduling",         # For ["scheduling", "rescheduling"]
    "SMS Confirmation",                    # For ["sms", "resent", "confirmation"]
    "Request for Extraordinary Reading",   # For ["reading", "extraordinary", "request"]
    "Scheduled Technical Visit",           # For ["scheduled", "campaign", "visit", "technical"]
    "Absent Technical Team",               # For ["technician", "not", "team", "appeared", "visit"]
    "Service Order Status"                 # For ["status", "service_order"]
]

Best regards,

**Reply by MaartenGr** [reply]
> Since the label is embedded, I assume that a precise topic label should yield better outcomes than a more general one. For example, should I use a label in a specific format like "sms"_resent_confirmation" or a more general title like "SMS Confirmation"?

It generally depends on the underlying embedding model but I would advise using the latter as that often generates more accurate representations.

**Reply by PDuarteLopes** [reply]
Thanks for your time. I just have one more question, and I think it might be related to an old bug. When I use Zero-Shot Topic Modeling, I get good clustering, but with seemingly mismatched representations.

I’m sorry that I'm working with a Portuguese dataset, but I think it’s easy to get my point across. As you can see in the following example, the topic names I give to my Zero-Shot model generate good clusters, but with mismatched cluster assignments. The keywords for each zero-shot topic appear in the representations, but not in the 'correct'/supposed ones. I might have cluster based on the topics given «, but not in a correct order. Do you know why this is happening?

![image_2024-11-13_171630201](https://github.com/user-attachments/assets/9db09886-47e5-43a7-9c25-6fee18b9c6a9)

Best regards,

**Reply by MaartenGr** [reply]
@PDuarteLopes This might have something to do with your BERTopic version as I believe this was an issue before. Could you try it with the latest version (0.16.4)?

**Reply by PDuarteLopes** [reply]
I upgraded my version and made some more tests, and unfortunately, the same "error" keeps happening. Even with different Zero-Shot names, the model cannot make cohesive associations between the Zero-Shot Topics and the respective Representations.

Going deeper, I started analyzing the cosine similarity between the embedding of each Zero-Shot Topic and my documents. I found out that even the documents with cosine similarity above 0.9 are being assigned to new topics during the training of my model. Why does this happen? Is it possible that the model initially attributes the document to my Zero-Shot topic and then finds a new topic with higher cosine similarity and makes the change? How can I avoid it?

I really see the potential of using Zero-Shot topics, but I am having a really hard time taking the full potential out of it. I am making a list of potential topics which are easily detected in my documents, but the model insists on creating new topics and giving non-logical representations to the given ones.

**Reply by MaartenGr** [reply]
Hmmm, that's strange! I can imagine this must be frustrating... 

I don't wont to make you do too much work but do you by change have a reproducible example for me? Maybe with some toy data? If I can reproduce what is happening, then I can start diving into this problem...

Also, if that's not possible. Could you share your code? That might help me understand what is happening here.
